lib/rouge/lexers/j.rb

```diff
@@ -0,0 +1,231 @@
+      end
+
+      def lex_str(match)
+        unless match[3]
+          token Error
+          return
+        end
+        token Str::Single, match[1]
+        match[2].scan(/([^']*)('*)/) do |s, e|
+          token Str::Single, s
+          token Str::Escape, e
+        end
+        token Str::Single, match[3]
+      end
+
+      state :expr do
+        rule %r/\s+/, Text
+
+        rule %r'([!-&(-/:-@\[-^`{-~]|[A-Za-z]\b)([.:]*)' do |m|
+          token J.primitive(m[1], m[2])
+        end
+
+        rule %r/(?:\d|_\d?):([.:]*)/ do |m|
+          token m[1].empty? ? J.token_map[:verb] : Error
+        end
+
+        rule %r/[\d_][\w.]*([.:]*)/ do |m|
+          token m[1].empty? ? Num : Error
+        end
+
+        rule(/(')((?:[^'\n]|'')*)(')?/) {|m| lex_str m }
+
+        rule %r/NB\.(?![.:]).*/, Comment::Single
+
+        rule %r/([A-Za-z]\w*)([.:]*)/ do |m|
+          if m[2] == '.'
+            word, sep, id = m[1].partition '_'
+            list = if sep.empty?
+              J.control_words
+            elsif not id.empty?
+              J.control_words_id
+            end
+            if list and list.include? word
+              token Keyword, word + sep
+              token((word == 'for' ? Name : Name::Label), id)
+              token Keyword, m[2]
+            else
+              token Error
+            end
+          else
+            token m[2].empty? ? Name : Error
+          end
+        end
+      end
+
+      start do
+        @q_expr ||= self.dup.tap {|o| o.push :q_expr }
+        @def_stack = []
+      end
+
+      state :root do
+        rule %r/\n/ do
+          token Text
+          @def_stack.empty? or @def_stack.each(&method(:push)).clear
+        end
+
+        # https://code.jsoftware.com/wiki/Vocabulary/com
+        # https://code.jsoftware.com/wiki/Vocabulary/NounExplicitDefinition
+        rule %r/
+          ([0-4]|13|adverb|conjunction|dyad|monad|noun|verb)([\ \t]+)
+          (def(?:ine)?\b|:)(?![.:])([\ \t]*)
+        /x do |m|
+          groups Keyword::Pseudo, Text, Keyword::Pseudo, Text
+          @def_body = (m[1] == '0' || m[1] == 'noun') ? :noun : :code
+          if m[3] == 'define'
+            @def_stack.push @def_body
+          else
+            push :expl_def
+          end
+        end
+
+        rule %r/^([ \t]*)(Note\b(?![.:]))([ \t\r]*)(?!=[.:]|$)/ do
+          groups Text, Name, Text
+          # 'Note' is a (pro)verb, and evaluated after applying modifiers
+          # including explicit definitions.
+          @def_stack.unshift :note
+        end
+
+        rule %r/[mnuvxy]\b(?![.:])/, Name
+        mixin :expr
+      end
+
+      state :expl_def do
+        rule %r/0\b(?![.:])/ do
+          token Keyword::Pseudo
+          @def_stack.push @def_body
+          pop!
+        end
+        rule %r/(')((?:[^'\n]|'')*)(')?/ do |m|
+          if @def_body == :noun
+            lex_str m
+          elsif m[3]
+            token Punctuation, m[1]
+            delegate @q_expr, m[2]
```

Comment:
I see what you're doing here - but I wonder if there's a way to accomplish this without `delegate` - as in, match just the first opening `'` (and maybe some lookahead), then push `:q_expr`, and trust it to find the closing delimiter and pop in the right place. Are there instances of `'` that don't trigger this state?

Reply:
I mention this because often for string literals and similar it's desirable to lex *within* them, both for syntax-embedding purposes and also to be able to highlight escapes and other special characters. I don't know J, so it's possible this isn't a concern either.